lib/raft.ts

```diff
@@ -0,0 +1,1381 @@
+          return;
+        }
+        if (this.state === 'Candidate') {
+          raftLog.info("ZWED0294I - Got final vote from %s and became leader of term %d", peerAddress, term);
+          done = true;
+          this.convertToLeader();
+        }
+      });
+    }
+    this.scheduleElectionOnTimeout();
+  }
+
+  convertToLeader(): void {
+    this.state = 'Leader';
+    // When a leader first comes to power, it initializes all nextIndex values to the index just after the last one in its log (11 in Figure 7)
+    const logLen = this.len();
+    for (let i = 0; i < this.peers.length; i++) {
+      this.nextIndex[i] = logLen;
+      this.matchIndex[i] = -1;
+    }
+    this.tracePrintf("nextIndex %s", JSON.stringify(this.nextIndex));
+    this.tracePrintf("matchIndex %s", JSON.stringify(this.matchIndex));
+    setImmediate(() => this.emitState());
+    this.sendHeartbeat();
+  }
+
+  private emitState(): void {
+    this.stateEmitter.emit('state', this.state);
+  }
+
+  sendHeartbeat(): void {
+    const peerCount = this.peers.length;
+
+    for (let server = 0; server < peerCount; server++) {
+      if (server == this.me) {
+        continue;
+      }
+      setImmediate(async () => {
+        if (!this.isLeader()) {
+          this.tracePrintf("cancel heartbeat to %d at term %d because not leader anymore", server, this.currentTerm);
+          return;
+        }
+        this.tracePrintf("sends heartbeat to %d at term %d", server, this.currentTerm);
+        const { ok, success, conflict } = await this.callAppendEntries(server, this.currentTerm, 'heartbeat');
+        if (ok && !success) {
+          if (this.isLeader() && conflict) {
+            this.tracePrintf("got unsuccessful heartbeat response from %d, adjust nextIndex because of conflict %s",
+              server, JSON.stringify(conflict));
+            this.adjustNextIndexForServer(server, conflict);
+          }
+        } else if (ok && success) {
+          this.tracePrintf("got successful heartbeat response from %d at term %d, nextIndex = %d, matchIndex = %d, commitIndex = %d",
+            server, this.currentTerm, this.nextIndex[server], this.matchIndex[server], this.commitIndex)
+          this.checkIfCommitted();
+        }
+      });
+    }
+    if (!this.isLeader()) {
+      this.tracePrintf("stop heartbeat because not leader anymore");
+      return;
+    }
+    this.heartbeatTimeoutId = setTimeout(() => this.sendHeartbeat(), this.heartbeatInterval)
+  }
+
+  private adjustNextIndexForServer(server: number, conflict: Conflict) {
+    if (conflict.conflictIndex === -1 && conflict.conflictTerm === -1) {
+      if (conflict.logLength === 0 && this.lastSnapshot) {
+        this.tracePrintf("follower's log is empty(have it re-started?) and there is a snapshot, send the snapshot to the follower");
+        setImmediate(() => this.installSnapshotForServer(server, this.currentTerm, this.lastSnapshot));
+      } else {
```

Comment:
It means the follower has just restarted so it has empty log, and the leader doesn't have a snapshot yet because there are not enough entries in the raft log. It this cases the leader will send log entries to the follower starting from beginning.